contracts/multisig-prover/src/encoding/sol/mod.rs

```diff
@@ -0,0 +1,115 @@
+use std::collections::BTreeMap;
+
+use axelar_encoding::types::{
+    CrossChainId, Message, Payload, PublicKey, Signature, Signer, WeightedSignature, WorkerSet,
+    U256,
+};
+
+use itertools::Itertools;
+use multisig::{msg::SignerWithSig, verifier_set::VerifierSet};
+
+const ECDSA_COMPRESSED_PUBKEY_LEN: usize = 33; // this should be probably a public constant in axelar_encoding.
+const ED25519_PUBKEY_LEN: usize = 32; // this should be probably a public constant in axelar_encoding.
+
+pub fn to_worker_set(vs: &VerifierSet) -> WorkerSet {
+    let signers: BTreeMap<String, Signer> = vs
+        .signers
+        .iter()
+        .map(|(address, signer)| {
+            let enc_signer = signer.address.to_string();
+            let enc_pubkey = to_pub_key(&signer.pub_key);
+
+            let enc_weight = U256::from_be(to_u256_be(signer.weight.u128()));
+
+            (
+                address.clone(),
+                Signer::new(enc_signer, enc_pubkey, enc_weight),
+            )
+        })
+        .collect();
+
+    WorkerSet::new(
+        vs.created_at,
+        signers,
+        U256::from_be(to_u256_be(vs.threshold.u128())),
+    )
+}
+
+fn to_pub_key(pk: &multisig::key::PublicKey) -> PublicKey {
+    match pk {
+        multisig::key::PublicKey::Ecdsa(hb) => {
+            PublicKey::new_ecdsa(hb.to_array::<ECDSA_COMPRESSED_PUBKEY_LEN>().unwrap())
+        }
+        multisig::key::PublicKey::Ed25519(hb) => {
+            PublicKey::new_ed25519(hb.to_array::<ED25519_PUBKEY_LEN>().unwrap())
+        }
+    }
+}
+
+// Fits a u128 into a u256 in big endian representation.
+fn to_u256_be(u: u128) -> [u8; 32] {
+    let mut uin256 = [0u8; 32];
+    uin256[16..32].copy_from_slice(&u.to_be_bytes());
+    uin256
+}
+
+impl From<&crate::payload::Payload> for Payload {
+    fn from(value: &crate::payload::Payload) -> Self {
+        match value {
+            crate::payload::Payload::Messages(msgs) => {
+                Payload::new_messages(msgs.iter().map(to_msg).collect_vec())
+            }
+            crate::payload::Payload::VerifierSet(vs) => Payload::new_worker_set(to_worker_set(&vs)),
+        }
+    }
+}
+
+fn to_msg(msg: &router_api::Message) -> Message {
+    let enc_cc_id = CrossChainId::new(msg.cc_id.chain.to_string(), msg.cc_id.id.to_string());
+
+    Message::new(
+        enc_cc_id,
+        msg.source_address.to_string(),
+        msg.destination_chain.to_string(),
+        msg.destination_address.to_string(),
```

Comment:
So all this fields, and it's related types, can be built again from a `String` representation. So obviously we are doing good here, as we are being agnostic. The decoding part would need know beforehand how to represent the decoded `String` on respective types. Where is that decoding part ? I am missing something.

Reply:
> The decoding part would need know beforehand how to represent the decoded String on respective types. Where is that decoding part ?

Maybe this is a fait point in that the `https://github.com/eigerco/solana-axelar-internal/blob/98a6ef5747f2398a42abfeb85bec8ab91b51cec8/solana/crates/axelar-encoding/src/types.rs#L102-L105` crate should operate on the original `router_api::Message` types rather than Strings?

I am looking at how the flow goes:
1. The relayer gets the raw bytes from amplifier API and tries to decode it using a hardcoded decoder [[link to code]](https://github.com/eigerco/solana-axelar-internal/blob/main/offchain/components/relayer/src/approver.rs?plain=1#L161) - in the linked example we're using bcs decoding (will be replaced with rkyv soon).
2. The messages get [bcs decoded](https://github.com/eigerco/solana-axelar-internal/blob/98a6ef5747f2398a42abfeb85bec8ab91b51cec8/solana/helpers/axelar-message-primitives/src/command/decoded_command.rs?plain=1#L111-L123) (this is our code btw). Looking at the implementation, most of the parameters are just left as bytes and we don't really interact with them. The only parameter that gets interpreted in any way is the `destination_address`.

When I was working on the "e2e" tests between evm and solana, I was relying on the [Pubkey::to_string()](https://github.com/eigerco/solana-axelar-internal/blob/98a6ef5747f2398a42abfeb85bec8ab91b51cec8/solana/programs/axelar-solana-memo-program/tests/evm-e2e/from_evm_to_solana.rs?plain=1#L105) implementation when being invoked on the source chain [(here's the decoding)](https://github.com/eigerco/solana-axelar-internal/blob/98a6ef5747f2398a42abfeb85bec8ab91b51cec8/solana/programs/axelar-solana-memo-program/tests/evm-e2e/from_evm_to_solana.rs?plain=1#L87). iirc Pukey::to_string() is a base58 encoded string.

Where is the issue: We force the origin chain to produce strings of "destination_address" that must be properly pared by the destination program. 

Solution: nothing we can really do here? Because theoretically its up to the caller on the source chain to properly encode `destination_address` so that we can parse it. Anyone can call a source gateway and pass destination addresses as hex encoded, binary encoded strings, base58 encoded strings, etc. And we cannot realistically handle this. 

And if we want this encoding be generic, then we cannot force it to use base58 encoded strings for ed25519 addresses either. This means that the parsing of the string gets propagated to the relayer.

Reply:
Are the different parts of a same flow (chain scoped, so integration scoped) able to know how to properly handle the data at each stage right ? If yes, i would probably leave it chain agnostic.

Reply:
> Well, since we are in control of the final types, we can force the destination_address to be [u8; 32], just like solana_program::pubkey::Pubkey. The fit will be even tighter.

The benefits of this:
- we can directly map the bytes into a public key (this only affects the relayer)
- there's no evm or cosmos chain compatibility that we'd have to worry about (they use 20 bytes for addresses)

On the other hand the interface for the EVM Gateway expects that the address [is already provided as a string rather than raw bytes](https://github.com/eigerco/solana-axelar-internal/blob/98a6ef5747f2398a42abfeb85bec8ab91b51cec8/evm-contracts/src/AxelarMemo.sol#L34-L43). 

iirc I tried doing exactly this (trying to send raw bytes, then converting them to a String inside Solidity, then sending it to the EVM gateway). But I had major issues with paring the events - the data didn't make sense. I think this is because ethers library generates the interfaces and types for events, and it expects the string parameter to be a valid UTF-8 string - but Solidity does not enforce strings to be utf8 hence the issue in parsing.

I think that we should rather leave the parameter as a String and keep the parsing logic on the relayer - with the assumption that the origin chain always sends base58 encoded to parse string as base58 for Solana. I think its easy to reason with this assumptoin. Even now, every gmp packet is expected to uphold some arbitrary encoding enforced by the destination relayer.